### ISSUE TITLE
Check constraint enforcing a column is NOT NULL does not satisfy missing_non_null_constraint

### ISSUE THREAD

**Issue by owst** [gregnavis/active_record_doctor#83]
**tl;dr**: `missing_non_null_constraint` gives a false positive when a column has a validated check constraint enforcing that the column is NOT NULL.
 
We use Postgres, and have added a (validated) [check constraint](https://www.postgresql.org/docs/11/ddl-constraints.html#DDL-CONSTRAINTS-CHECK-CONSTRAINTS) that ensures there are no NULL values in a particular column, but we're seeing a false positive from `missing_non_null_constraint`. We used a check constraint in place of a NOT NULL constraint since the table in question is large, and we didn't want to hold the `ACCESS EXCLUSIVE` lock taken by `SET NOT NULL` for the duration of scanning the whole table to assert that the column contains no NULL values. Adding a `NOT VALID` constraint is very quick (no lengthy table scan) and `VALIDATE CONSTRAINT` only takes a `SHARE UPDATE EXCLUSIVE` lock, which doesn't block concurrent read/writes to the table. Interestingly, it sounds like with Postgres 12 (which we're not yet using), if there is a validated `CHECK CONSTRAINT` enforcing NOT NULL on a column, `SET NOT NULL` can be performed without the expensive table scan as it can rely on the validated check constraint.

N.b. A check constraint can assert an arbitrary boolean expression, so it may be slightly tricky to determine if a given constraint always enforces NOT NULL, but I think for basic constraints it should be straightforward.

Here's a reproduction script, which assumes a PG DB has been created using `createdb my_test_db` and then the script is run as `DB_NAME=my_test_db DB_USERNAME=... DB_PASSWORD=... ruby repro.rb`:

```ruby
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails", "= 5.2.6.2"
  gem "pg", "= 1.3.2"
  gem "active_record_doctor", "= 1.9.0"
end

require "active_record"
require "active_record_doctor"

ActiveRecord::Base.establish_connection(
  adapter: "postgresql",
  database: ENV["DB_NAME"],
  username: ENV["DB_USERNAME"],
  password: ENV["DB_PASSWORD"]
)

# ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Base.connection.execute(<<~SQL)
  DROP TABLE IF EXISTS things;

  CREATE TABLE things(name text, number integer NOT NULL);

  -- Ensure any newly created rows have a name
  ALTER TABLE things ADD CONSTRAINT name_not_null CHECK (name IS NOT NULL) NOT VALID;

  -- Ensure all existing rows conform to the constraint (typically performed separately due to the
  -- differing locks that are taken by both operations)
  ALTER TABLE things VALIDATE CONSTRAINT name_not_null;
SQL

class Thing < ActiveRecord::Base
  validates_presence_of :name
  validates_presence_of :number
end

runner = ActiveRecordDoctor::Runner.new(ActiveRecordDoctor.load_config_with_defaults(nil))
runner.run_one(:missing_non_null_constraint)

Thing.create!(name: "present", number: 1)

thing = Thing.new(name: nil, number: 2)

puts "Thing without name is valid?: #{thing.valid?}"

begin
  # Ignore the validation and save anyway - it should blow up
  thing.save!(validate: false)
rescue => e
  puts "Caught #{e.message}"
end
```

The relevant output:
```
add `NOT NULL` to things.name - models validates its presence but it's not non-NULL in the database
Thing without name is valid?: false
Caught PG::CheckViolation: ERROR:  new row for relation "things" violates check constraint "name_not_null"
DETAIL:  Failing row contains (null, 2).
: INSERT INTO "things" ("number") VALUES ($1)
```
Showing that there is a presence validation and that the DB prevents null values for `name`, but there is a false positive.

**Comment by fatkodima** [comment]
@owst This is a very valid request due to the reasons about locking and large tables you mentioned.

I solved the similar problem recently:
https://github.com/fatkodima/online_migrations/blob/99cb0c2e2594e8fc638a341f0cf028a4c946f19f/lib/online_migrations/command_checker.rb#L302-L309

I will create a PR.

**Comment by gregnavis** [comment]
@owst, thank you for such **fantastic** bug report. Kudos for the effort put into writing it. 🙇🏻‍♂️

@fatkodima already produced a PR which I'll have a look at shortly. Thank you for such quick turnaround time, @fatkodima!

I'd like to notice, we should probably add a mirror check to `missing_presence_validator` and make it require a validator if there's an appropriate check constraint. What do you guys think?

**Comment by fatkodima** [comment]
I think this is a great idea. I forgot about it. Will update my PR.
